sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbGridFSIO.java

```diff
@@ -0,0 +1,427 @@
+        }
+      };
+    }
+  }
+
+  /** Read data from GridFS. */
+  public static Read<String> read() {
+    return new Read<String>(new Read.BoundedGridFSSource<String>(null, null, null, null,
+                            StringsParseCallback.INSTANCE, StringUtf8Coder.of()));
+  }
+
+  static class Read<T> extends PTransform<PBegin, PCollection<T>> {
+    public Read<T> withUri(String uri) {
+      return new Read<T>(new BoundedGridFSSource<T>(uri, options.database,
+                                           options.bucket, options.filterJson,
+                                           options.parser, options.coder));
+    }
+
+    public Read<T> withDatabase(String database) {
+      return new Read<T>(new BoundedGridFSSource<T>(options.uri, database,
+                                           options.bucket, options.filterJson,
+                                           options.parser, options.coder));
+    }
+
+    public Read<T> withBucket(String bucket) {
+      return new Read<T>(new BoundedGridFSSource<T>(options.uri, options.database, bucket,
+          options.filterJson, options.parser, options.coder));
+    }
+
+    public <X> Read<X> withParsingFn(ParseCallback<X> f) {
+      return new Read<X>(new BoundedGridFSSource<X>(options.uri, options.database,
+          options.bucket, options.filterJson, f, null));
+    }
+
+    public Read<T> withCoder(Coder<T> coder) {
+      return new Read<T>(new BoundedGridFSSource<T>(options.uri, options.database,
+          options.bucket, options.filterJson, options.parser, coder));
+    }
+
+    public Read<T> withQueryFilter(String filterJson) {
+      return new Read<T>(new BoundedGridFSSource<T>(options.uri, options.database,
+          options.bucket, filterJson, options.parser, options.coder));
+    }
+
+    private final BoundedGridFSSource<T> options;
+
+    Read(BoundedGridFSSource<T> options) {
+      this.options = options;
+    }
+
+    @Override
+    public PCollection<T> apply(PBegin input) {
+      org.apache.beam.sdk.io.Read.Bounded<T> unbounded =
+          org.apache.beam.sdk.io.Read.from(options);
+      PCollection<T> output = input.getPipeline().apply(unbounded);
+      if (options.coder != null) {
+        output.setCoder(options.coder);
+      }
+      return output;
+    }
+
+    static class BoundedGridFSSource<T> extends BoundedSource<T> {
+      @Nullable
+      private final String uri;
+      @Nullable
+      private final String database;
+      @Nullable
+      private final String bucket;
+      @Nullable
+      private final String filterJson;
+      @Nullable
+      private final ParseCallback<T> parser;
+      @Nullable
+      private final Coder<T> coder;
+      @Nullable
+      private List<ObjectId> objectIds;
+      private transient Mongo mongo;
+      private transient GridFS gridfs;
+
+      BoundedGridFSSource(String uri, String database, String bucket, String filterJson,
+          ParseCallback<T> parser, Coder<T> coder) {
```

Comment:
Are both of these constructors necessary? Seems like the caller can simply call the second constructor and pass null as objectIds.